terraform/provider/resource_teleport_role.go

```diff
@@ -1,61 +1,145 @@
-		Create: resourceTeleportRoleUpsert,
-		Read:   resourceTeleportRoleRead,
-		Update: resourceTeleportRoleUpsert,
-		Delete: resourceTeleportRoleDelete,
-		Exists: resourceTeleportRoleExists,
+		CreateContext: resourceRoleCreate,
+		ReadContext:   resourceRoleRead,
+		UpdateContext: resourceRoleUpdate,
+		DeleteContext: resourceRoleDelete,
 
 		Schema: tfschema.SchemaRoleV3(),
 	}
 }
 
-func resourceTeleportRoleUpsert(d *schema.ResourceData, m interface{}) error {
-	// client := m.(*client.Client)
-	name := d.Get("name").(string)
+// resourceRoleCreate creates Teleport role
+func resourceRoleCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	c, err := getClient(m)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
 
-	// roleSpec := services.RoleSpecV3{}
+	name, err := getRoleName(d)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
 
-	// role, err := services.NewRole(name, roleSpec)
-	// if err != nil {
-	// 	return trace.Wrap(err)
-	// }
+	role, err := types.NewRole(name, types.RoleSpecV3{})
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
 
-	// FIXME: Client.CreateRole() is not implemented yet in api/client
-	// err = client.CreateRole(role)
-	// if err != nil {
-	// 	return trace.Wrap(err)
-	// }
+	roleV3, ok := role.(*types.RoleV3)
+	if !ok {
+		return diag.FromErr(trace.Errorf("failed to convert created role to types.RoleV3"))
+	}
 
-	d.SetId(name)
-	return nil
+	err = tfschema.UnmarshalRoleV3(d, roleV3)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
+
+	err = c.UpsertRole(ctx, roleV3)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
+
+	d.SetId(roleV3.GetName())
+
+	return resourceRoleRead(ctx, d, m)
 }
 
-func resourceTeleportRoleRead(d *schema.ResourceData, m interface{}) error {
-	// client := m.(*auth.Client)
-	// name := d.Get("name").(string)
+// resourceRoleRead reads Teleport role
+func resourceRoleRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	c, err := getClient(m)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
 
-	return nil
+	roleID := d.Id()
+
+	_, err = c.GetRole(ctx, roleID)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
+
+	return diag.Diagnostics{}
 }
 
-func resourceTeleportRoleDelete(d *schema.ResourceData, m interface{}) error {
-	// client := m.(*auth.Client)
-	// name := d.Get("name").(string)
+// resourceRoleUpdate updates Teleport role
+func resourceRoleUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	c, err := getClient(m)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
 
-	return nil
+	roleID := d.Id()
+
+	role, err := c.GetRole(ctx, roleID)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
+
+	roleV3, ok := role.(*types.RoleV3)
+	if !ok {
+		return diag.FromErr(trace.Errorf("failed to convert created role to types.RoleV3"))
+	}
+
+	err = tfschema.UnmarshalRoleV3(d, roleV3)
```

Comment:
Does it update the retrieved `roleV3` from the contents of `d`? If so, add a comment clarifying this please.

Reply:
No, it does not. This is the existence check. It is required because we do not have separate update method for Role (it's UpsertRole), and we do not need to create non-existent Role here, even if Terraform thinks it exists.

Reply:
Added clarifying comment